### PR TITLE
[SPARK-6145][SQL] fix ORDER BY on nested fields

### PR DESCRIPTION
**Pull request by marmbrus**
This PR is based on work by @cloud-fan in #4904, but with two differences:
- We isolate the logic for Sort's special handling into `ResolveSortReferences`
- We avoid creating UnresolvedGetField expressions during resolution.  Instead we either resolve GetField or we return None.  This avoids us going down the wrong path early on.

### CODE REVIEW

**Comment by cloud-fan** [comment]
Since we decide to resolve `Sort` in `ResolveSortReferences`, then we need to stop resolve `Sort` in `ResolveReferences`. For example,

    test("quick") {
        jsonRDD(sparkContext.makeRDD(
          """{"a": {"b": 1}, "b": 2}""" ::
          """{"a": {"b": 2}, "b": 1}""" :: Nil)).registerTempTable("t")
        checkAnswer(sql("SELECT a.b FROM t ORDER BY b"), Seq(Row(2), Row(1)))
      }
This test will fail, because we have mistakenly resolved `b` as `a.b` in `ResolveReferences`(we try to resolve `b` on top of `Projection("a.b")` and success). A quick fix is just make sure we don't resolve `Sort` in `ResolveReferences` like `case u @ UnresolvedAttribute(name) if !q.isInstanceOf[Sort]`. Do you have better idea @marmbrus ?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 1]
```diff
@@ -340,19 +340,16 @@ class Analyzer(catalog: Catalog,
...
-          Project(projectList.map(_.toAttribute),
-            Sort(ordering, global,
-              Project(projectList ++ missingInProject, child)))
+          Project(p.output,
+            Sort(resolvedOrdering, global,
+              Project(projectList ++ missing, child)))
         } else {
-          logDebug(s"Failed to find $missingInProject in ${p.output.mkString(", ")}")
+          logDebug(s"Failed to find $missing in ${p.output.mkString(", ")}")
           s // Nothing we can do here. Return original plan.
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Even there is no missing in project, we still need to build new `Sort` with resolved `SortOrder`.

**Comment by cloud-fan** [comment]
Another problem, for something like `SELECT DISTINCT a FROM t ORDER BY a`, we can't resolve this `Sort` because its child is not `Project` or `Aggregate`, thus can't handled by `ResolveSortReferences`.

**Comment by marmbrus** [comment]
@cloud-fan, thanks for the further comments.  Regarding not resolving sort in in `ResolveReferences`, I agree that is a little confusing and I think you are correct for standard SQL.  However, this is not how hive does resolution and I don't think that we can differ from their implementation.  Consider the following:

```
  test("hive orderby queries") {
    val df = jsonRDD(sparkContext.makeRDD(
      """{"a": {"b": 1}, "b": 2}""" ::
      """{"a": {"b": 2}, "b": 1}""" :: Nil))registerTempTable("t")
    sql("CREATE TABLE t1 AS SELECT * FROM t")
    runSqlHive("SELECT a.b FROM t1 ORDER BY b").foreach(println)
    sql("SELECT a.b FROM t1 ORDER BY b").collect().foreach(println)
  }
```

Which outputs:

```
1
2
[1]
[2]
```

So given this, I think the current way we resolve things (i.e. first resolve based on what comes from the select clause and then fall back on its child only when this fails) is non-standard but what we must do to avoid breaking existing HiveQL queries.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala [hunk 2]
```diff
@@ -196,14 +197,19 @@ abstract class LogicalPlan extends QueryPlan[LogicalPlan] with Logging {
...
-        val fieldExprs = nestedFields.foldLeft(a: Expression)(UnresolvedGetField)
-        val aliasName = nestedFields.last
-        Some(Alias(fieldExprs, aliasName)())
+        try {
+
+          // The foldLeft adds UnresolvedGetField for every remaining parts of the name,
+          // and aliased it with the last part of the name.
+          // For example, consider name "a.b.c", where "a" is resolved to an existing attribute.
+          // Then this will add UnresolvedGetField("b") and UnresolvedGetField("c"), and alias
+          // the final expression as "c".
```
Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala]: Update comment? Seems we will not add `UnresolvedGetField` any more.

**Comment by cloud-fan** [comment]
Ah, so `SELECT a.b` is equal to `SELECT a.b AS b`, and the `b` in `ORDER BY b` should reference to `a.b`, that's how hive does resolution and makes sense. I have sent you a PR to remove the duplicated `resolveGetField` and others LGTM.

**Comment by yhuai** [comment]
LGTM

**Pull request closed without merging**